File: algorithms/301_RemoveInvalidParentheses.py
class Solution:
    def removeInvalidParentheses(self, s):
        left = 0
        right = 0

        for char in s:
            if char == '(':
                left += 1
            elif char == ')':
                right = right + 1 if left == 0 else right
                left = left - 1 if left > 0 else left

        result = {}
        def recurse(s, index, left_count, right_count, left_rem, right_rem, expr):
            if index == len(s):
                if left_rem == 0 and right_rem == 0:
                    ans = "".join(expr)
                    result[ans] = 1
            else:
                if (s[index] == '(' and left_rem > 0) or (s[index] == ')' and right_rem > 0):
                    recurse(s, index + 1,
                            left_count,
                            right_count,
                            left_rem - (s[index] == '('),
                            right_rem - (s[index] == ')'), expr)

                expr.append(s[index])
                if s[index] != '(' and s[index] != ')':
                    recurse(s, index + 1,
                            left_count,
                            right_count,
                            left_rem,
                            right_rem, expr)
                elif s[index] == '(':
                    recurse(s, index + 1,
                            left_count + 1,
                            right_count,
                            left_rem,
                            right_rem, expr)
                elif s[index] == ')' and left_count > right_count:
                    recurse(s, index + 1,
                            left_count,
                            right_count + 1,
                            left_rem,
                            right_rem, expr)
                expr.pop()                 

        recurse(s, 0, 0, 0, left, right, [])     
        return list(result.keys())

File: algorithms/301_RemoveInvalidParentheses.java
class Solution {
    public List<String> removeInvalidParentheses(String s) {
        /**
         * This solution should work for all cases when no other characters are involved.
         */
        return helper(s, false);
    }
    
     public List<String> helper(String s, boolean reverse) {
        List<String> result = new LinkedList<String>();
        String validParentheses = "";
        int left = 0, right = 0;
        char temp, open='(', close=')';
        if (reverse) {
            open = ')';
            close = '(';
        }
         
        List<Integer> cBP = new ArrayList<Integer>();
        
        for (int i=0; i<s.length();++i) {
            temp = s.charAt(i);
            if (reverse)
                validParentheses = temp + validParentheses;
            else
                validParentheses += temp;
            if (temp != close) {
                if (temp == open)
                    ++left;
            } else {
                ++right;
                cBP.add(i);
                if (right > left) {
                    Set<String> set = new HashSet<String>();
                    for (int j=0;j<cBP.size();++j) {
                        if (!reverse)
                            set.add(validParentheses.substring(0, cBP.get(j)) +
                                    validParentheses.substring(cBP.get(j)+1));
                        else
                            set.add(validParentheses.substring(0, i-cBP.get(j)) +
                                validParentheses.substring(i-cBP.get(j)+1));
                    }
                        
                    
                    List<String> validParaOfRemaining;
                    if (!reverse)
                        validParaOfRemaining = helper(s.substring(i+1), reverse);
                    else
                        validParaOfRemaining = helper(s.substring(i+1), reverse);
                        
                    Iterator<String> setIt = set.iterator(), remainingIt;
                    while (setIt.hasNext()) {
                        String a = setIt.next();
                        remainingIt = validParaOfRemaining.iterator();
                        while (remainingIt.hasNext())
                            result.add(a + remainingIt.next());
                    }
                    break;
                }
            }
        }
        
        if (result.size() == 0) {
            if (left == right)
                return new ArrayList<String>(Arrays.asList(validParentheses));
            byte [] strAsByteArray = s.getBytes(); 
            byte [] reverseByteArray = new byte [strAsByteArray.length]; 
  
            for (int i = 0; i<strAsByteArray.length; i++) 
                reverseByteArray[i] =  strAsByteArray[strAsByteArray.length-i-1];
            return helper(new String(reverseByteArray), !reverse);
        }
        
        return result;
    }
}